moveit_py/src/moveit_py/pybind11_utils/include/moveit_py/pybind11_utils/ros_msg_typecasters.h

```diff
@@ -0,0 +1,124 @@
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * The name of Robert Haschke may not be use to endorse or promote
+ *     products derived from this software without specific prior
+ *     written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Peter David Fagan, Robert Haschke */
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp/serialization.hpp>
+#include <rclcpp/duration.hpp>
+
+namespace py = pybind11;
+
+namespace pybind11
+{
+namespace detail
+{
+
+// Base class for type conversion (C++ <-> python) of ROS message types
+template <typename T>
+struct RosMsgTypeCaster
+{
+  // C++ -> Python
+  // TODO: add error handling
+  static handle cast(const T& src, return_value_policy /* policy */, handle /* parent */)
+  {
+    // serialize src
+    rclcpp::Serialization<T> serializer;
+    rclcpp::SerializedMessage serialized_msg;
+    serializer.serialize_message(&src, &serialized_msg);
+    py::bytes bytes = py::bytes(reinterpret_cast<const char*>(serialized_msg.get_rcl_serialized_message().buffer),
+                                serialized_msg.get_rcl_serialized_message().buffer_length);
+
+    // get Python object type
+    const std::string ros_msg_name = rosidl_generator_traits::data_type<T>();
+
+    // find delimiting '::' in ros_msg_name
+    std::size_t pos1 = ros_msg_name.find("::");
+    std::size_t pos2 = ros_msg_name.find("::", pos1 + 2);
+    py::module m = py::module::import((ros_msg_name.substr(0, pos1) + ".msg").c_str());
+
+    // retrieve type instance
+    py::object cls = m.attr(ros_msg_name.substr(pos2 + 2).c_str());
+
+    // deserialize into python object
+    py::module rclpy = py::module::import("rclpy.serialization");
+    py::object msg = rclpy.attr("deserialize_message")(bytes, cls);
+
+    return msg.release();
+  }
+
+  // Python -> C++
+  bool load(handle src, bool /*convert*/)
+  {
+    // serialize src into python buffer
+    py::module rclpy = py::module::import("rclpy.serialization");
+    py::bytes bytes = rclpy.attr("serialize_message")(src);
```

Comment:
You need to ensure that `src` has the expected type:
```cpp
if (src.attr("__class__") != pyclass_from_name(rosidl_generator_traits::name<T>())
  return false;
```
... otherwise the serialized message (`bytes`) will belong to a wrong type.

Reply:
[58a541f](https://github.com/ros-planning/moveit2/pull/1546/commits/58a541fbf78c0f492de267af15597b03ee181ed5)